plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1860,6 +1858,161 @@ public Optional<ConnectorOutputMetadata> finishInsert(
+    {
+        if (commitSummary.getProtocol().isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Protocol changed by concurrent write operation");
+        }
+    }
+
+    private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
+    {
+        if (!commitSummary.getMetadataUpdates().isEmpty()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Metadata changed by concurrent write operation");
+        }
+    }
+
+    private static void checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain<DeltaLakeColumnHandle> enforcedPartitionConstraint, IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<AddFileEntry> addedFiles = switch (isolationLevel) {
+            case WRITESERIALIZABLE -> commitSummary.getIsBlindAppend().orElse(false)
+                    ? List.of() // Do not conflict with blind appends
+                    : commitSummary.getAddedFiles();
+            case SERIALIZABLE -> commitSummary.getAddedFiles();
+        };
+
+        if (addedFiles.isEmpty()) {
+            return;
+        }
+
+        boolean readWholeTable = enforcedPartitionConstraint.isAll();
+        if (readWholeTable) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files added in the modified table by concurrent write operation.");
+        }
+
+        Map<DeltaLakeColumnHandle, Domain> enforcedDomains = enforcedPartitionConstraint.getDomains().orElseThrow();
+        Optional<AddFileEntry> unreadAddAction = addedFiles.stream()
+                .filter(addAction -> partitionMatchesPredicate(addAction.getCanonicalPartitionValues(), enforcedDomains))
+                .findFirst();
+        if (unreadAddAction.isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Data files were  added on the table with isolation level %s in the partition '%s' of the modified table  by concurrent write operation."
+                    .formatted(isolationLevel, unreadAddAction.get().getPartitionValues()));
+        }
+    }
+
+    private static void checkIfCommittedRemovedFilesConflictWithCurrentOperation(IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<RemoveFileEntry> removedFiles = switch (isolationLevel) {
+            case WRITESERIALIZABLE -> commitSummary.getIsBlindAppend().orElse(false)
+                    ? List.of() // Do not conflict with blind appends
+                    : commitSummary.getRemovedFiles();
+            case SERIALIZABLE -> commitSummary.getRemovedFiles();
+        };
+
+        if (removedFiles.isEmpty()) {
+            return;
+        }
+
+        // TODO Record add files read per table to figure out whether the removed files do actually conflict with the read table partitions
```

Comment:
We can do much better here.

`remove` action should contain the partition values as well  which would help figuring out (in the same fashion as for the `add` files whether we are reading from partitions removed by a previously commited winning operation.

https://github.com/delta-io/delta/blob/13739bf7740a9a2bee11d4e46dcab6e52d4fa8f3/connectors/standalone/src/main/java/io/delta/standalone/actions/RemoveFile.java#L70-L75